flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTask.java

```diff
@@ -195,6 +195,22 @@ protected void createInputProcessor(
         // EndOfPartitionEvent, we would not complement barriers for the
         // unfinished network inputs, and the checkpoint would be triggered
         // after received all the EndOfPartitionEvent.
+        if (options.getCheckpointType().shouldDrain()) {
+            CompletableFuture<Void> sourcesStopped =
+                    FutureUtils.waitForAll(
+                            operatorChain.getSourceTaskInputs().stream()
+                                    .map(s -> s.getOperator().stop())
+                                    .collect(Collectors.toList()));
+
+            return sourcesStopped.thenCompose(
+                    ignore -> triggerSourcesCheckpointInMailbox(metadata, options));
```

Comment:
Don't we need something like `FutureUtils.assertNoException` wrapped around this? In case of an exception happening after completing `sourcesStopped`, but before enqueuing mailbox action?

ditto about other places when we are composing futures?

Reply:
Isn't `thenCompose` working similarly as `flatMap`, meaning that if either `sourcesStopped` or enqueing the mailbox action fails with an exception, the combined future will be completed exceptionally and will fail the RPC trigger?

Reply:
Forget about, what I said. The result of the future is ignored in the Task. You're right we need that :(

Reply:
Yes & no. Task doesn't wait on the `CompletableFuture` returned from `triggerCheckpointAsync` the. As far as I remember the future is used currently only in tests (it has been used long time ago in production code as well I think), but we kind of stuck with it because we didn't know how to refactor tests to not use it (I'm open to discussions to get rid of this effectively test only code).

But that's kind of beside the point here, because the production code is ignoring if the future fails or not.

Reply:
Sorry, if I am asking dumb question...

After looking at it again, I am not sure if we need to do anything about it. The future from `sourceOperator#stop()` never completes exceptionally. It can only ever complete successfully if the `END_OF_DATA` was emitted. At the same time the `StreamTask#triggerCheckpointAsync` submits a mailbox action as the very first thing.

Do we still need any additional handling there?

Reply:
Yes. We have learnt it the hard way in other places in the code, that all exceptions, even the least likely should be handled and should bubble up. Including NPE, type castings or other clearly unexpected errors. Without `assertNoException` tests/code can fail in a weird way or deadlock for example, obscuring the root problem (some trivial NPE). On top of that submitting mail to mailbox can fail as well with `RejectedExecutionException`. 

Speaking of.  With `assertNoException`, `RejectedExecutionException` should probably be explicitly ignored here? If task was cancelled in the mean time, before we managed to submit mailbox action, I think in the old code `Task` would ignore this exception because of `CANCELLING` state. In new one we should probably do the same thing.